### PR TITLE
Add support for Dynatrace metrics API v2

### LINKED ISSUE #24978: Add auto-configuration for Micrometer's Dynatrace v2 meter registry
See related Issue & PR :  
     https://github.com/micrometer-metrics/micrometer/issues/2295
     https://github.com/micrometer-metrics/micrometer/pull/2406

With the hopeful addition of the new Micrometer Dynatrace v2 Meter Registry, it will be necessary to add an additional Auto Configuration for it to this project in order to utilize it. This will allow users to add in the necessary application properties needed for dynatrace2 to send metrics to Dynatrace. 

I do have a working example for this, but cannot submit a PR as the micrometer PR is still open (as of 2021/22/1) and the required dependency would cause the build to fail.

### PR DESCRIPTION
**Pull request by pirgeo**
This PR contains new additions to the properties that allow Micrometer to export to two different Dynatrace APIs. The code in the two Micrometer PRs (see below) depends on this, and will hopefully be included in the 1.8.0 release of Micrometer.

Resolves #24978

- Micrometer issue: https://github.com/micrometer-metrics/micrometer/issues/2295
- Move Dynatrace exporter v1: https://github.com/micrometer-metrics/micrometer/pull/2553
- Add Dynatrace exporter v2: https://github.com/micrometer-metrics/micrometer/pull/2581

### CODE REVIEW

**Comment by snicoll** [comment]
@pirgeo thank you for the PR but I am going to close this now as it isn't actionable. There is no way for us to review the  code as the changes in Micrometer are not available yet. When micrometer has a build with the necessary changes of this PR, we can obviously reconsider this.

**Comment by jonatan-ivanov** [comment]
@snicoll The changes were merged to micrometer: https://github.com/micrometer-metrics/micrometer/pull/2619

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/dynatrace/DynatraceProperties.java [hunk 1]
```diff
@@ -97,4 +139,36 @@ public void setGroup(String group) {
...
+
+	public Boolean getEnrichWithOneAgentMetadata() {
+		return enrichWithOneAgentMetadata;
+	}
+
+	public void setEnrichWithOneAgentMetadata(Boolean enrichWithOneAgentMetadata) {
+		this.enrichWithOneAgentMetadata = enrichWithOneAgentMetadata;
+	}
+
+	public Map<String, String> getDefaultDimensions() {
```
Comment (jonatan-ivanov) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/dynatrace/DynatraceProperties.java:158]: I'm not sure if it is the preferred way in spring boot, but I might make this immutable: `Collections.unmodifiableMap(defaultDimensions);` (I don't really remember if configprops will like it or not).

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/dynatrace/DynatracePropertiesConfigAdapter.java [hunk 2]
```diff
@@ -16,14 +16,17 @@
...
-
 import org.springframework.boot.actuate.autoconfigure.metrics.export.properties.StepRegistryPropertiesConfigAdapter;
 
+import java.util.Map;
+
 /**
  * Adapter to convert {@link DynatraceProperties} to a {@link DynatraceConfig}.
  *
  * @author Andy Wilkinson
+ * @author Georg Pirklbauer
```
Comment (jonatan-ivanov) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/dynatrace/DynatracePropertiesConfigAdapter.java:29]: I think you can add yourself to the previous file and to the test class too.

File: spring-boot-project/spring-boot-docs/src/docs/asciidoc/production-ready-features.adoc [hunk 3]
```diff
@@ -1723,25 +1723,87 @@ You can also change the interval at which metrics are sent to Datadog:
...
+Please ensure that the endpoint URI contains the path (e.g. `/api/v2/metrics/ingest`).
+
+[source,yaml,indent=0,configprops,configblocks]
+----
+	management:
+	  metrics:
+	    export:
+	      dynatrace:
+	        api-version: "v2"
+	        api-token: "YOUR_TOKEN" # should be read from environment variables or properties
```
Comment (jonatan-ivanov) [inline spring-boot-project/spring-boot-docs/src/docs/asciidoc/production-ready-features.adoc:1758]: I would just say from a secure property source (there are other occurrences in this file).

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/dynatrace/DynatraceProperties.java (hunk 1 above)
Comment (wilkinsona) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/dynatrace/DynatraceProperties.java:160]: It should be modifiable for consistency with other configuration properties classes.

File: spring-boot-project/spring-boot-docs/src/docs/asciidoc/production-ready-features.adoc [hunk 4]
```diff
@@ -1723,25 +1723,87 @@ You can also change the interval at which metrics are sent to Datadog:
...
-	        api-token: "YOUR_TOKEN"
 	        device-id: "YOUR_DEVICE_ID"
-	        uri: "YOUR_URI"
+	        # uri: "https://{your-domain}/e/{your-environment-id}" on managed deployments.
+	        uri: "https://{your-environment-id}.live.dynatrace.com"
+	        api-token: "YOUR_TOKEN" # should be read from a secure property source
 ----
 
-You can also change the interval at which metrics are sent to Dynatrace:
+You can also change the interval at which metrics are sent to Dynatrace (works for both API versions).
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-docs/src/docs/asciidoc/production-ready-features.adoc:1805]: As this applies to both API versions, I don't think it belongs in the API v1 section.

Comment (pirgeo) [reply spring-boot-project/spring-boot-docs/src/docs/asciidoc/production-ready-features.adoc:1807]: You are right, I added a subsection for version independent settings.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/dynatrace/DynatraceProperties.java (hunk 1 above)
Comment (pirgeo) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/dynatrace/DynatraceProperties.java:160]: I reset it to be modifiable. Thanks for the clarification!

File: spring-boot-project/spring-boot-docs/src/docs/asciidoc/production-ready-features.adoc (hunk 3 above)
Comment (pirgeo) [reply spring-boot-project/spring-boot-docs/src/docs/asciidoc/production-ready-features.adoc]: Changed it everywhere.

**Comment by wilkinsona** [comment]
Given that there are some configuration properties that are specific to v1 and some that are specific to v2, I wonder if we should introduce separate groups for those. Properties that are common to both version of the API would remain at `management.metrics.export.dynatrace` while v1-specific properties would move to `management.metrics.export.dynatrace.v1` and, similarly, v2-specific properties would move to `management.metrics.export.dynatrace.v2`. WDYT, @snicoll?

**Comment by snicoll** [comment]
I think that's an interesting idea. That said our influx support has the same problem and we're currently mapping the same way Micrometer does. I don't know if there's a precedent where our config diverges quite a bit from the related Micrometer's `Config` class.

**Comment by pirgeo** [comment]
Would that be a breaking change, or is there a way to do this without breaking existing `v1` configurations? 
@jonatan-ivanov Have you come across a similar situation and have input here?

**Comment by jonatan-ivanov** [comment]
@pirgeo On Boot side it depends how the Boot Team wants to handle this.
If we can come up with a worthy breaking change for Micrometer, that can be the subject of 2.x.
As Stéphane mentioned the Influx config is similar where the v1-v2 is handled similarly to this PR, though I really like Andy's proposal.

**Comment by pirgeo** [comment]
Thanks @jonatan-ivanov. I also think the proposal makes sense. We would also prefer to introduce this as a non-breaking, additive change. This would allow us to provide a version compatible with Micrometer 1.x without having to wait for a 2.x release.
@snicoll, @wilkinsona Would the current proposal in this PR be fine with you?

I have also moved the documentation to the new location. The CI is currently breaking, since this code relies on the `1.8.0-SNAPSHOT` build of Micrometer. How can we go ahead here?

**Comment by jonatan-ivanov** [comment]
@snicoll @wilkinsona Would it make sense to you to depend on Micrometer 1.8.0 in `main` early on so that Boot support for new features in Micrometer could be merged in?

**Comment by wilkinsona** [comment]
Yes, I think so. We can do that once we have some dates for the Micrometer 1.8 milestones and we know that the schedules line up.

**Comment by jonatan-ivanov** [comment]
@wilkinsona We are planning to release M1/M2/M3/RC1/GA 8 days before Boot releases these milestones, e.g.: `1.8.0-M1` is planned on 14th July, `2.6.0-M1` is on 22nd July. Can we include this in `2.6.0-M1`?

**Comment by pirgeo** [comment]
I updated the PR to reflect the renaming of the config option introduced in https://github.com/micrometer-metrics/micrometer/pull/2692

**Comment by wilkinsona** [comment]
I've pushed [a branch](https://github.com/wilkinsona/spring-boot/tree/gh-26258) which hopefully illustrates what I had in mind. The approach is to move v1- and v2-specific properties beneath `management.metrics.export.dynatrace.v1` and `management.metrics.export.dynatrace.v2` respectively. For backwards compatibility, the old v1-specific properties remain at `management.metrics.export.dynatrace` in a deprecated form.

Rather than having a property for the API version, it is now inferred based on whether or not the device ID has set. A downside of this approach is that a user who is attempting to use the V1 API will not get an error message is they forget to set the device ID. Instead, an attempt will be made to use the V2 API. The upside of the approach is that the IDE experience is better as auto-completion guides you towards the right properties and there's no chance of, for example, setting the API version to v1 and then trying to configure v2-specific properties.

**Comment by pirgeo** [comment]
Hi @wilkinsona, Thanks for the suggestion, this is great. We have a small change in mind: would it be possible to rename `v1` and `v2` to `api-v1`  and `api-v2`, respectively? `v1`/`v2` sounds like registry-versions similar to `@since` where features can be mixed. When using `api-v1` or `api-v2` it's more clear that we mean the Dynatrace API version, which should clarify that only one of the two (the latest one by default) is used, and features cannot be mixed between versions. What do you think? I have enabled "Allow maintainers to modify" on this PR, so you should be able to push your commit to the open branch if you'd like, otherwise I can take that over. I'll put some finishing touches the docs after we've settled for one of the alternatives
Thanks!

**Comment by wilkinsona** [comment]
Thanks for taking a look, @pirgeo. We don't use hyphens in configuration property prefixes so we try to keep each `.`-separated section of the prefix to a single word. We could use `apiv1` and `apiv2` or `v1api` and `v2api` but not `api-v1` and `api-v2`. Personally I prefer the concision of `v1` and `v2` and don't find it confusing, but you know your user base far better than I do. For `v1api` vs `apiv1` I prefer the former as the latter could be read as apiv 1.

**Comment by pirgeo** [comment]
@wilkinsona Oh, I see. `apiv1` and `apiv2` does indeed look a bit odd. Lets go ahead with `v1` and `v2` then, and I will update the documentation to make it very clear for everyone reading it. Would you push your commit to this branch, or do you want me to do that? Thanks again for the support!

**Comment by wilkinsona** [comment]
GitHub's giving me a 403 when I try to push to the `add-dynatrace-v2-properties` branch of `dynatrace-oss-contrib/spring-boot`. Please feel free to push my commit onto the branch yourself.

**Comment by pirgeo** [comment]
@wilkinsona  I have pushed your commit and refined the documentation, it looks good from our end now. Thanks for your help!
I think the build will still break due to this branch pointing to Micrometer 1.7.0, which doesn't have our changes. How should we proceed here?

**Comment by wilkinsona** [comment]
I can take care of that. I'll rebase it before merging it and that'll pick up the upgrade to 1.8.0-M1.

**Comment by wilkinsona** [comment]
@pirgeo Thanks very much for making your first contribution to Spring Boot. This has now been merged into `main`.

**Comment by pirgeo** [comment]
@wilkinsona Thanks a lot!

**Pull request closed without merging**